core/src/main/java/com/linecorp/armeria/common/DefaultTimeoutController.java

```diff
@@ -0,0 +1,183 @@
+            final long passedTimeMillis = TimeUnit.NANOSECONDS.toMillis(currentNanoTime - lastStartTimeNanos);
+            final long newTimeoutMillis = timeoutMillis + adjustmentMillis - passedTimeMillis;
+            timeoutMillis = newTimeoutMillis;
+            lastStartTimeNanos = currentNanoTime;
+            if (newTimeoutMillis > 0) {
+                timeoutFuture = eventLoopSupplier.get().schedule(
+                        timeoutTask, newTimeoutMillis, TimeUnit.MILLISECONDS);
+            } else {
+                // We went past the dead line set by the new timeout already.
+                timeoutTask.onTimeout();
+            }
+        }
+    }
+
+    @Override
+    public void resetTimeout(long newTimeoutMillis) {
+        ensureInitialized();
+        if (newTimeoutMillis <= 0) {
+            timeoutMillis = newTimeoutMillis;
+            cancelTimeout();
+            return;
+        }
+
+        if (timeoutTask.canReschedule()) {
+            final long currentNanoTime = System.nanoTime();
+            final long passedTimeMillis = TimeUnit.NANOSECONDS.toMillis(currentNanoTime - lastStartTimeNanos);
+            final long remainingTimeoutMillis = timeoutMillis - passedTimeMillis;
+            lastStartTimeNanos = currentNanoTime;
+            if (remainingTimeoutMillis == newTimeoutMillis) {
+                return;
+            }
+
+            // Cancel the previously scheduled timeout, if exists.
+            cancelTimeout();
+            timeoutMillis = newTimeoutMillis;
+            timeoutFuture = eventLoopSupplier.get().schedule(timeoutTask, newTimeoutMillis,
+                                                             TimeUnit.MILLISECONDS);
+        }
+    }
+
+    @Override
+    public boolean cancelTimeout() {
+        final ScheduledFuture<?> timeoutFuture = this.timeoutFuture;
+        if (timeoutFuture == null) {
+            return true;
+        }
+
+        this.timeoutFuture = null;
+        return timeoutFuture.cancel(false);
+    }
+
+    private void ensureInitialized() {
+        checkState(firstStartTimeNanos > 0,
+                   "initTimeout(timeoutMillis) is not called yet.");
+    }
+
+    @Override
+    public long startTimeNanos() {
+        return firstStartTimeNanos;
+    }
+
+    @VisibleForTesting
+    long timeoutMillis() {
+        return timeoutMillis;
+    }
+
+    @Nullable
+    @VisibleForTesting
+    ScheduledFuture<?> timeoutFuture() {
+        return timeoutFuture;
+    }
+
+    /**
+     * A timeout task that is invoked when the deadline exceeded.
+     */
+    public interface TimeoutTask extends Runnable {
+        /**
+         * Returns {@code true} if the timeout task is ready to start.
+         */
+        boolean isReady();
+
+        /**
+         * Returns {@code true} if the timeout task can be rescheduled.
+         */
+        boolean canReschedule();
+
+        /**
+         * Invoked when the deadline exceeded.
+         */
+        void onTimeout();
```

Comment:
Having two methods for this seems weird to me. How about just using `run()` and adding Javadoc for `run()`?

Reply:
That sounds good to me. :-) Let me remove this, and use `run()`.